design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+# lakectl Testing
+
+## Purpose
+
+This document describes the design of a testing framework aimed at testing our CLI, namely `lakectl`.
+It specifies the requirements from the system and a set of technical questions we need to answer. Each question (in fact, each option), is accompanied by pros and cons, to help us understand which options will answer our needs the best. These are obviously not the only aspects to these decisions, and so, this document and the decisions derived from, are subject to reviews and suggestions.
+This document does not specify the cli tests to run, but the system intended to run them
+
+## Requirements
+* Usable both locally and as part of CI/CD
+* Ease of use - adding new tests should be as easy as possible. Moreover, adding new tests, or modifying existing ones, should not draw modifications to the testing framework itself
+  * It is obvious that changing the test **format** (e.g. expanding) will require framework modifications, in order to support the modified format
+* Decoupling of Framework and Data - the test data format should contain all the required info for the test itself, regardless of the framework design. In other words, we would like the ability to rewrite/replace the framework and still use the same tests data
+*
+
+## What We Want To Test 
+
+The goal is to the test `lakectl` correctness. At the very least we would like to verify that:
+* CLI input is parsed correctly
+* The correct requests are sent to the server (if any)
+* Server result is formatted correctly
+* Both erroneous input and error results from the server are formatted correctly
+* CLI Internal errors are handled correctly
+
+This testing framework is not designed to perform end-to-end tests (i.e. server correctness) or performance, for that matter
+
+## Architecture
+
+### Architecture Outline
+
+![lakectl testing architecture](./diagrams/lakectl-testing-arch.png)
+
+### Test Case Structure
+
+The test case structure should contain at least the following details:
+* pre-condition: the system state right before the tested command starts. This can be a function name to call, a mock type to use or a file containing lakectl commands to run in order to bring the system to that state (in case real server is in use)
+* lakectl command: a string with the `lakectl` command to test.
+* expected result: the expected output of a correct execution of the above command. This can be a string or, better, a file containing the output string
```

Comment:
Commented by @itaiad200 
I think that the expected result should also contain the request(s) that were sent to the server. I know it depends on the architecture that is described below, but many of the lakectl commands have no output at all, and other may display the correct output without performing the right actions.

Reply:
That is an important insight!
I'm trying to think how can we perform that. May be the testing framework should act as a proxy between `lakectl` to `lakefs` server/mock? That way we can verify the request contents and delegate it forwards to the recipient?
@itaiad200 WDYT?